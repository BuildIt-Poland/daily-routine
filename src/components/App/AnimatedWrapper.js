import React, { useEffect, useState } from 'react';
import { node } from 'prop-types';
import styled, { withTheme } from 'styled-components';
import { animated, useTransition, config } from 'react-spring';

import { location, theme } from '../../types';
import extractRoleFromPath from '../../utils/extractRoleFromPath';
import generateColors from '../../utils/generateColors';

const Wrapper = styled(animated.div)`
  position: absolute;
  width: 100%;
  min-height: 100%;
  display: grid;
  grid-template-rows: auto 1fr auto;
`;

function usePreviousRole(currentRole) {
  const [previousRole, setPreviousRole] = useState(currentRole);
  useEffect(() => setPreviousRole(currentRole), [currentRole]);

  return previousRole;
}

function AnimatedWrapper({ children, location, theme }) {
  const { pathname } = location;
  const { primaryColor } = theme;

  const currentRole = extractRoleFromPath(pathname);
  const previousRole = usePreviousRole(currentRole);

  const didRouteChanged = previousRole !== currentRole;
  const previousTheme = generateColors(previousRole);
  const previousPrimaryColor = previousTheme.primaryColor;

  const bgTransitions = useTransition(pathname, currentRole, {
    from: { backgroundColor: didRouteChanged ? previousPrimaryColor : primaryColor },
    enter: { backgroundColor: primaryColor },
    leave: { backgroundColor: primaryColor },
    config: config.gentle
  });

  return bgTransitions.map(
    ({ item, props, key }) =>
      item && (
        <Wrapper key={key} style={props}>
          {children}
        </Wrapper>
      )
  );
}

AnimatedWrapper.propTypes = {
  children: node.isRequired,
  location: location.isRequired,
  theme: theme.isRequired
};

export default withTheme(AnimatedWrapper);
