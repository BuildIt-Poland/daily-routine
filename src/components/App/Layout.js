import styled from 'styled-components';

const Layout = styled.div`
  min-height: 100%;
  display: grid;
  grid-template-rows: auto 1fr auto;
  background-color: ${props => props.theme.primaryColor};
`;

export default Layout;
