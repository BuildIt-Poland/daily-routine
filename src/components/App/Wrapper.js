import React, { useEffect, useState } from 'react';
import styled, { withTheme } from 'styled-components';
import { animated, useTransition, config } from 'react-spring';
import { colorLightTeal } from '../../styles/designTokens';
import extractRoleFromPath from '../../utils/extractRoleFromPath';
import generateColors from '../../utils/generateColors';

const Wrapper = styled(animated.div)`
  position: absolute;
  width: 100%;
  min-height: 100%;
  display: grid;
  grid-template-rows: auto 1fr auto;
`;

function Animator({ children, location, theme }) {
  const { pathname } = location;
  const { primaryColor } = theme;
  const currentRole = extractRoleFromPath(pathname);

  const [previousRole, setPreviousRole] = useState(currentRole);
  useEffect(() => setPreviousRole(currentRole), [previousRole, currentRole]);

  const didRouteChanged = previousRole !== currentRole;
  const { primaryColor: previousPrimaryColor } = generateColors(previousRole);

  const bgTransitions = useTransition(pathname, currentRole, {
    from: { backgroundColor: didRouteChanged ? previousPrimaryColor : primaryColor },
    enter: { backgroundColor: primaryColor },
    leave: { backgroundColor: primaryColor },
    config: config.wobbly
  });

  return bgTransitions.map(
    ({ item, props, key }) =>
      item && (
        <Wrapper key={key} style={props}>
          {children}
        </Wrapper>
      )
  );
}

export default withTheme(Animator);
