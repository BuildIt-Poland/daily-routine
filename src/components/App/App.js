import React from 'react';
import { ThemeProvider } from 'styled-components';
import { Route, withRouter, Switch } from 'react-router-dom';

import { FRONT_END_ROLE, BACK_END_ROLE, DEV_OPS_ROLE } from '../../constants/roles';
import { ROOT_PATH } from '../../constants/routes';
import { location, history } from '../../types';
import { GenderProvider } from '../../context/GenderContext';
import { RouteTransitionAnimationProvider } from '../../context/RouteTransitionAnimationContext';
import generateTheme from '../../utils/generateTheme';
import AppBar from '../AppBar';
import Footer from '../Footer';
import LandingPage from '../LandingPage';
import FrontEndRolePage from '../FrontEndRolePage';
import BackEndRolePage from '../BackEndRolePage';
import DevOpsRolePage from '../DevOpsRolePage';
import RouteTransitionAnimation from '../RouteTransitionAnimation';
import PageNotFound from '../PageNotFound';
import Wrapper from './Wrapper';

function App({ location, history }) {
  return (
    <RouteTransitionAnimationProvider history={history}>
      <ThemeProvider theme={generateTheme(location.pathname)}>
        <>
          <RouteTransitionAnimation location={location} />
          <GenderProvider>
            <Wrapper>
              <AppBar />
              <main>
                <Switch>
                  <Route exact path={ROOT_PATH} component={LandingPage} />
                  <Route path={`/${FRONT_END_ROLE}`} component={FrontEndRolePage} />
                  <Route path={`/${BACK_END_ROLE}`} component={BackEndRolePage} />
                  <Route path={`/${DEV_OPS_ROLE}`} component={DevOpsRolePage} />
                  <Route path="*" component={PageNotFound} />
                </Switch>
              </main>
              <Footer />
            </Wrapper>
          </GenderProvider>
        </>
      </ThemeProvider>
    </RouteTransitionAnimationProvider>
  );
}

App.propTypes = {
  location,
  history
};

export default withRouter(App);
