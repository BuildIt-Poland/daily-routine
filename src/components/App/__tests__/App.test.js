import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { create } from 'react-test-renderer';

import App from '../App';

jest.mock('../../WideButton', () => 'WideButton');
jest.mock('../../AppBar', () => 'AppBar');
jest.mock('../../Footer', () => 'Footer');
jest.mock('../../Characters', () => ({
  SugarCat: 'SugarCat'
}));

describe('COMPONENT - App', () => {
  it('renders correctly default route', () => {
    const component = create(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
