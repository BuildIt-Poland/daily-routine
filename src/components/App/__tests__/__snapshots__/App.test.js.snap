// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - App renders correctly default route 1`] = `
.c0 {
  min-height: 100%;
  display: grid;
  grid-template-rows: auto 1fr auto;
  background-color: #80cbd6;
}

.c1 {
  position: relative;
  padding: 2rem;
}

.c2 {
  font-size: 2.4rem;
  margin: 0;
  margin-bottom: 1rem;
}

.c3 {
  font-size: 1.8rem;
  margin: 1rem 0;
  font-weight: 400;
}

<div
  className="c0"
>
  <AppBar />
  <main>
    <article>
      <section
        className="c1"
      >
        <h1
          className="c2"
        >
          Don't know what to say?
        </h1>
        <h2
          className="c3"
        >
          Worry not! Find your perfect scenario for incoming daily.
        </h2>
        <SugarCat />
      </section>
      <section
        className="c1"
      >
        <h2
          className="c3"
        >
          Pick your role!
        </h2>
        <nav>
          <RoleButton
            variant="frontend"
          />
          <RoleButton
            variant="backend"
          />
          <RoleButton
            variant="devops"
          />
        </nav>
      </section>
    </article>
  </main>
  <Footer />
</div>
`;
